### ISSUE TITLE
Binary releases for Linux on ARM

### ISSUE THREAD

**Issue by mvglasow** [bcpierce00/unison#443]
As I use Unison to sync files between an x86_64-based laptop (running Ubuntu) and a Raspberry Pi-based file server, I would really appreciate a binary build for ARM platforms so I can use compatible versions on both ends.

Gold standards, of course, would be .deb packages (even better in the form of a PPA). Some folks might need RPM instead. Since I see you guys already have CI/CD set up, that should not be overly difficult, given the minimal dependencies of Unison.

**Comment by gdt** [comment]
As I see it, the biggest issue in CI is lack of operating system diversity - no BSDs, no Illumos.  Next on the list is CPU types.

However, my understanding is that the current CI setup more or less spans what github provides, and improvement awaits a pull request to use a different CI system or to add platforms that somehow github provides that  are not used.  Really, packaging systems for operating systems should be up to date, rather than each project somehow being obligated to build for 1000s of OS/version/CPU/ocaml versions.

**Comment by gdt** [comment]
See also #445, whcih is the big-picture parent of this issue.

**Comment by mvglasow** [comment]
quick one, does the toolchain support cross-compiling for ARM? If so, I could give ARM support a shot.

ps: yeah, in a perfect world OS distributions would always have the latest version of upstream. In practice, though, they grab the then-current version once per release (if at all), and if you decide to take the route of stability and stick to the LTS versions, you get to enjoy the bugs for two years, until the next LTS replaces them with new ones. And it’s even more fun when you have one machine running Ubuntu, another one running Raspbian, both having different release cycles, and they need to be compatible somehow.</rant> But addressing this is probably an even bigger one than #445. Dunno how to solve this—for Unison, which has hardly any run-time dependencies, building packages for the major platforms out there should be feasible and cover all but the most exotic ones.

**Comment by gdt** [comment]
Please join unison-hackers at https://lists.seas.upenn.edu/mailman/listinfo/unison-hackers and ask about cross.
We are trying to have the issue tracker only be about bugs (and I don't know the answer to your question, but someone else might).

**Comment by mvglasow** [comment]
I did some research on my own. The OCaml compiler does not seem to support cross-compiling, so your best bet is to bootstrap a chroot jail with the necessary binaries for a minimal system for the target arch, then use QEMU with user mode emulation to install build dependencies and do the actual build.

Steps are:

```
export CROSS_ARCH=armhf
export CROSS_ROOT=/opt/chroot/$CROSS_ARCH
export DISTRO=stretch
export CROSS_MIRROR=http://deb.debian.org/debian/
apt-get update
apt-get -y install debootstrap qemu-user-static binfmt-support
mkdir -p $CROSS_ROOT
debootstrap --variant=buildd --include=fakeroot,build-essential --arch=$CROSS_ARCH --foreign $DISTRO $CROSS_ROOT $CROSS_MIRROR
mkdir -p $CROSS_ROOT/usr/bin
cp /usr/bin/qemu-arm-static $CROSS_ROOT/usr/bin/
chroot $CROSS_ROOT ./debootstrap/debootstrap --second-stage
```
This should work for any architecture supported by QEMU user mode, and any version (not sure how well this works cross-distribution, e.g. Debian/armhf on top of Ubuntu/amd64). Any commands that should run on the target architecture would then be invoked as:

    chroot $CROSS_ROOT qemu-arm-static /bin/bash -c "COMMAND"

I am still trying to figure out why I need to provide the full path to `bash` (or any other binary) despite it being on `$PATH` (and yes, I have checked `$PATH` is set correctly), or if I need to wrap the command in `bash` at all.

Once that is in place, we could start enhancing the CI script to produce cross-platform builds.

**Comment by gdt** [comment]
Thanks for the thoughts.

It's an interesting question about that vs providing a builder using qemu.  I am quite leery of cranking up the CI resource requirements to involve running qemu.   It never occurred to me to run qemu in a CI builder.

github CI has been more and more troubled, and we are contemplating the ejection button.  I wonder if you know about CI on codeberg, as it might be easier there.

**Comment by mvglasow** [comment]
No idea about Codeberg, I keep my own projects on GitLab – if you guys opt for GitLab, I could help with that.

I have never dug down deeply into github CI, but after some research, I realize its constraints – mostly that it is limited to a handful of images to run the build on. GitLab lets you use any Docker image from the usual repositories, including your own. Though you still don’t get to pick the hardware architecture – unless you set up and maintain your own runners.

If we decide to start cross-platform CI on GHA and migrate elsewhere later, we’ll want to keep things portable – that is, rely as little as possible on the features of a specific CI. I would suggest creating one job per architecture, and doing all the setup through OS commands – that should be easy to plug into a different CI.

Re full path to `bash`, I am told this is normal behavior – the QEMU binary is not a shell and does not do `PATH` parsing. So you either need to provide the full path to your binary, or wrap it in `bash -c`, with the full path to `bash`.

**Comment by mvglasow** [comment]
Update: according to https://docs.codeberg.org/ci/, Codeberg uses Woodpecker CI, but so far that seems to be available to select projects only.

**Comment by mvglasow** [comment]
In the meantime I came across a tutorial to cross-compile Unison using QEMU. It’s slightly different from what I attempted, as it sets up a full VM with a disk image (which takes time but needs to be done only once):

https://gael-varoquaux.info/misc/unison-248-binaries-for-arm.html

This was tested with Unison 2.48, back in 2016. They rely on Debian 8 (jessie), which no longer seems to work as it is no longer available from the repositories I tried. Installing Debian 9 (stretch) seems to work, which is the version I need to support in my own environment. Tests are still ongoing.

**Comment by mvglasow** [comment]
Update: I was able to build v2.52.1 from git with `make UISTYLE=text NATIVE=true STATIC=true`. As `ldd ./src/unison` tells me, it is a static binary with no dependencies regarding libraries on the system.

As a nice side effect, the static armhf binary also appears to run on Android (`unison -version` prints its version number), both on 32-bit and 64-bit arm architectures.

Due to the old OCaml version (4.02.3) on this setup, v2.53.0 and higher no longer build – that would take a later OCaml version or a distribution upgrade. So, as a stepping stone, build 2.52.1 against the OCaml version you need and stick with that until you have upgraded your archives and all peers, then build 2.53.0 against a modern OCaml version.

**Comment by gdt** [comment]
For what it's worth, I can tell you that building unison on a RPI3 under NetBSD (machine type earmv7hf-el) works fine:
```
unison version 2.53.0 (ocaml 4.14.0)
```

As far as I can tell, the sources are fine and one just needs a working build environment.

**Comment by mvglasow** [comment]
Partial success with cross-platform CI, see https://github.com/mvglasow/unison/tree/cross-ci. (Beware, the history is really dirty, and likely to seeing some forced push operations.)

Basically, I am setting up a foreign chroot environment for the target platform with `debootstrap` and make some modifications to replicate the CI environment inside the chroot (basically, remounting `/proc` and hardlinking a few dirs).

Steps that run in the foreign chroot, i.e. on the target architecture, use a custom shell, which I also supply. That custom shell is basically a wrapper script which starts a bash shell in the foreign chroot, sets a few environment variables and the working dir, then runs the script for the step.

It was necessary to run `opam init` with `--disable-sandboxing`. Without that, `opam switch` fails when running `bwrap` due to lack of user namespace support in the kernel – this may be a limitation of the VM setup.

The CLI build already completes without errors, although I haven’t been able to examine the artifacts yet.

Currently there are still two issues:

* Building the GUI fails at the step where the artifact is being copied. (The step was copied from the `build` job with no changes, other than running it in a custom shell.) I am still investigating what is going on here – possibly the wrapper script has trouble dealing with the filename building magic. At the worst, I’d have to hardcode the binary name – since this job is Linux-only, some corner cases (related to `.exe` suffixes and similar) do not apply here.
* Setting up OCaml takes over an hour. `opam init` completes in two minutes. Currently we’re 15 minutes into `opam switch`, after which `opam install opam-depext` will run. I am still investigating which step takes the bulk of the time.

**Comment by mvglasow** [comment]
The copying issue was due to the `PROJECT_EXES` environment variable not being propagated to the foreign chroot. Adding that to the bash wrapper script fixed that. This leaves us with the opam performance issue.

**Comment by tleedjarv** [comment]
You've already done a lot of work based on the current GHA setup, so maybe you don't want to throw this effort away, but one advice I can offer you is to start fresh. The `PROJECT_EXES`, for example, and the whole `Initialize workflow variables` step and related steps is very heavy. Majority of it is needed for Windows builds and for publishing build results during a release. Further, some of it tries to be a very generic CI script that works with any project. All in all, this is an overkill in my view, especially with GHA (which automatically means there's nothing generic about it). Starting fresh and going step by step, you'd probably end up replicating some similar bits but probably much lighter.

I'd recommend getting the most simple build working – it will be very-very simple (I mean, just look at some of the builds in the current workflow, like docs and the bytecode build) – next add the GUI build, then publishing the build results and only then think about making the script more generic. It would be easier to debug this way as well.

> Setting up OCaml takes over an hour. `opam init` completes in two minutes. The next command, `opam switch create`, stays stuck in the first `Processing actions` stage for about an hour,

This is definitely not normal. With the resources provided by GHA, this step should take maybe 10 minutes without a cache.

**Comment by mvglasow** [comment]
Basically, I just copied the existing `build` job into a new one and adapted the steps, inserting some to set up a foreign chroot environment and modifying the build steps to run in that environment (which is just a matter of changing the shell). The majority of the work was to get the foreign chroot working, and to provide a ‘wrapper’ shell which passes all commands to a shell in the target environment – and these things should be fairly portable.

While I don’t want to promise too much – if I find some time, I might start a PoC on Gitlab where I will slim things down. One thing I have in mind is to create Docker images with the necessary build environment. All the build environment setup then goes into the Dockerfile, whereas the CI script does little more than build the code and publish it. There are still some unknowns in this: Gitlab CI does not have matrices, so we’ll need to replace them into something else. Also, I have no experience with non-Linux platforms on Gitlab – that can be anything from easy-peasy to impossible.

**Comment by tleedjarv** [comment]
> Basically, I just copied the existing `build` job into a new one and adapted the steps, inserting some to set up a foreign chroot environment and modifying the build steps to run in that environment (which is just a matter of changing the shell). The majority of the work was to get the foreign chroot working, and to provide a ‘wrapper’ shell which passes all commands to a shell in the target environment – and these things should be fairly portable.

If that's fine with you then no problem. I was thinking that maybe you need to do it in context of #858. The jobs are very complex (mostly due to the nature of GHA), a non-GHA CI script could be much simpler.
